webrtc/README.md

```diff
@@ -0,0 +1,412 @@
+  based on it?
+
+## Multiplexing
+
+The WebRTC browser APIs do not support half-closing of streams nor resets of the
+sending part of streams.
+[`RTCDataChannel.close()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel/close)
+flushes the remaining messages and closes the local write and read side. After
+calling `RTCDataChannel.close()` one can no longer read from the channel. This
+lack of functionality is problematic, given that libp2p protocols running on top
+of transport protocols, like WebRTC, expect to be able to half-close or reset a
+stream. See [Connection Establishment in
+libp2p](https://github.com/libp2p/specs/blob/master/connections/README.md#definitions).
+
+To support half-closing and resets of streams, libp2p WebRTC uses message
+framing. Messages on a `RTCDataChannel` are embedded into the Protobuf message
+below and sent on the `RTCDataChannel` prefixed with the message length in
+bytes, encoded as an unsigned variable length integer as defined by the
+[multiformats unsigned-varint spec][uvarint-spec].
+
+It is an adaptation from the [QUIC RFC]. When in doubt on the semantics of
+these messages, consult the [QUIC RFC].
+
+``` proto
+syntax = "proto2";
+
+package webrtc.pb;
+
+message Message {
+  enum Flag {
+    // The sender will no longer send messages on the stream.
+    FIN = 0;
+    // The sender will no longer read messages on the stream. Incoming data is
+    // being discarded on receipt.
+    STOP_SENDING = 1;
+    // The sender abruptly terminates the sending part of the stream. The
+    // receiver can discard any data that it already received on that stream.
+    RESET_STREAM = 2;
+  }
+
+  optional Flag flag=1;
+
+  optional bytes message = 2;
+}
+```
+
+Note that "a STOP_SENDING frame requests that the receiving endpoint send a
+RESET_STREAM frame.". See [QUIC RFC - 3.5 Solicited State
+Transitions](https://www.rfc-editor.org/rfc/rfc9000.html#section-3.5).
+
+Encoded messages including their length prefix MUST NOT exceed 16kiB to support
+all major browsers. See ["Understanding message size
+limits"](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Using_data_channels#understanding_message_size_limits).
+Implementations MAY choose to send smaller messages, e.g. to reduce delays
+sending _flagged_ messages.
+
+### Ordering
+
+Implementations MAY expose an unordered byte stream abstraction to the user by
+overriding the default value of `ordered` `true` to `false` when creating a new
+data channel via
+[`RTCPeerConnection.createDataChannel`](https://www.w3.org/TR/webrtc/#dom-peerconnection-createdatachannel).
+
+## Connection Security
+
+Note that the below uses the message framing described in
+[multiplexing](#multiplexing).
+
+While WebRTC offers confidentiality and integrity via TLS, one still needs to
+authenticate the remote peer by its libp2p identity.
+
+After [Connection Establishment](#connection-establishment):
+
+1. _A_ opens a WebRTC datachannel.
```

Comment:
the current Rust impl does things a little bit differently. A and B both open a channel with ID=1 and `negotiated` option set to `true` (https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/createDataChannel#parameters). 

the alternative is to wait for the first data channel on the server (B) and perform noise after data channel is negotiated. But I'm not sure this is strictly better than having a fixed ID=1 for that channel because 

a) we don't need to negotiate anything 
b) having a "special" ID underscores it's special purpose 
c) it may be that client opens multiple streams at once and server may pick "wrong" substream for noise.

what do people think?

Reply:
Is this ID typically auto-generated or how does this behave with future channels?

Reply:
> Is this ID typically auto-generated or how does this behave with future channels?

Typically, yes it's auto-generated.

Reply:
Okay, I like the idea in general but I think there are a few questions around how we detect a misbehaving peer. With the current spec, it is extremely crucial that we do not accept / use any other data channels before the handshake is finished.

This idea actually helps with that because both sides can delay registering a listener and only need to use the `createDataChannel` API in the beginning. If we adopt this, I think we should add a "Both sides MUST NOT accept incoming data channels until the handshake is complete."

- Does `createDataChannel` fail automatically if they other party doesn't open the same channel within X amount of time?
- What if they both specify a different ID? Does that trigger a failure?
- The spec says IDs start at 0. Why not use that?
- Should we pick a more special number like 25519? That may be an easier to understand constant in the code.
- We should double check that it doesn't matter if a channel with the same ID appears later.

Reply:
> Does createDataChannel fail automatically if they other party doesn't open the same channel within X amount of time?

No, it will not fail. If the channel is negotiated, the side that created the channel can send messages and will view the channel's `readyState` as 'open',  but the receiving side will only receive any buffered messages once they also create the channel. If the receive buffer is full incoming messages are discarded.

> What if they both specify a different ID? Does that trigger a failure?

No, in that case if negotiated is set to true, they both see `readyState = open`, and can send messages, but the other side will never receive. It will be buffered to the corresponding id and once the buffer is full it is discarded. 

> The spec says IDs start at 0. Why not use that?
Should we pick a more special number like 25519? That may be an easier to understand constant in the code.

I'm not sure the choice of constant matters. We picked 1, so we could just treat that as the special constant and document it. 

> We should double check that it doesn't matter if a channel with the same ID appears later.
It does matter. On the browser, if you create a channel with label 'test' and id 5, and then create another channel 'test1' with id 5, libwebrtc will automatically assign a different id for the channel with label 'test1'. We should use a fixed constant for the handshake channel and let the implementations auto-select id's for subsequent channels.

Reply:
> A and B both open a channel with ID=1 and `negotiated` option set to `true`

This sounds good to me. Thanks for raising this @melekes.

_Not a strong opinion, thus a small suggestion to avoid confusion:_

We would only set the `negotiated` option for the first special Datachannel used for the Noise handshake. For the consecutive datachannels we would not set `negotiated` and thus the user agent chooses a channel ID for us. What does Firefox and Chrome do? Do they choose `0`, `2` or some random number? I would find it confusing if the first (Noise) Datachannel is `1` and the second `0` and would thus advocate for the Noise channel to use `0`.

Reply:
They will choose the number sequentially as far as I have observed.

Reply:
Thus when using `1` for the Noise data channel, the browser would choose `2` for the consecutive data channel, correct @ckousik?

Reply:
> > Does createDataChannel fail automatically if they other party doesn't open the same channel within X amount of time?
> 
> No, it will not fail. If the channel is negotiated, the side that created the channel can send messages and will view the channel's `readyState` as 'open', but the receiving side will only receive any buffered messages once they also create the channel. If the receive buffer is full incoming messages are discarded.

Okay this is good to know. Not sure how relevant for the spec but we definitely need timeouts here in the implementations then.

> > The spec says IDs start at 0. Why not use that?
> > Should we pick a more special number like 25519? That may be an easier to understand constant in the code.
> 
> I'm not sure the choice of constant matters. We picked 1, so we could just treat that as the special constant and document it.

Similar to what @mxinden suggests, I think we should pick 0 so that they are definitely consecutive.

> > If we adopt this, I think we should add a "Both sides MUST NOT accept incoming data channels until the handshake is complete."
> 
> Not sure how strict we should make this. Consider that the initiator (client/browser) leads the receiver in the connection establishment process. During the Go implementation, I have run into a race condition where the client is able to set up the datachannel callbacks and initiate a stream before the server is able to set up it's callbacks. The solution was to set up callbacks before the handshake, and then allow the handshake to happen. Even if we do not set up the callbacks, the underlying implementations will create the datachannel and buffer any incoming data.

This race condition is exactly what I think we have to avoid. Until the authentication process is complete, the other data channels cannot be considered secure. As far as I understand, the nice thing about `negotiated: true` is that we don't need to set up callbacks because of parties can use the `createDataChannel` API which will return the data channel to use for negotiating. Together with a timeout in case of a misbehaving peer, this should give us certainty that we are not processing data until the handshake is finished.

Reply:
> Thus when using `1` for the Noise data channel, the browser would choose `2` for the consecutive data channel, correct @ckousik?

correct

Reply:
note: the reason for not picking `0` in Rust implementation was a bug in webrtc-rs, which has been fixed at some point. So we can switch to `0`.

Reply:
It might be easier to make ourselves independent from WebRTC datachannel IDs here. For example, we might decide that we want to have another bespoke stream in addition (or replacing?) the handshake stream we're defining now, and we won't be able to do that if we define that ID `X` is the handshake stream and all other streams are normal libp2p streams.

We had a very similar problem in HTTP/3, where there are different types of unidirectional streams (and in the short time since the publication of RFC 9114 we've already regretted not using a similar design for bidirectional design). The solution we came up with was defining a "stream type", which is simply a varint sent at the beginning of every (unidirectional) stream. For details see https://datatracker.ietf.org/doc/html/rfc9114#section-6.2.

Reply:
WebRTC's data channels already have a ["protocol"](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/createDataChannel#parameters). We could use that maybe?

Reply:
That looks like a promising idea, I like it! We could leave it empty for normal libp2p streams, and only set it to something non-empty for our handshake stream.

Reply:
> This race condition is exactly what I think we have to avoid. Until the authentication process is complete, the other data channels cannot be considered secure. As far as I understand, the nice thing about `negotiated: true` is that we don't need to set up callbacks because of parties can use the `createDataChannel` API which will return the data channel to use for negotiating. Together with a timeout in case of a misbehaving peer, this should give us certainty that we are not processing data until the handshake is finished.

@thomaseizinger we can process the datachannels only after the handshake succeeds, but setting the callbacks after the handshake is complete causes the  race condition.

Reply:
> This race condition is exactly what I think we have to avoid. Until the authentication process is complete, the other data channels cannot be considered secure. As far as I understand, the nice thing about negotiated: true is that we don't need to set up callbacks because of parties can use the createDataChannel API which will return the data channel to use for negotiating. Together with a timeout in case of a misbehaving peer, this should give us certainty that we are not processing data until the handshake is finished.

There are legitimate use cases for creating streams and sending data before authenticating the peer. Obviously, you need to be careful with this (how "being careful" looks like will very much depend on your programming language's API). For example, the server might want to send the list of supported protocols via Identify (see the discussion in #454). Not sure if that's the way we'd go for that particular example (using the Noise Extension Registry would be another option), but we should keep us that option open.

In particular that means: On the receiver side we don't block accepting streams on handshake completion.

Reply:
> > This race condition is exactly what I think we have to avoid. Until the authentication process is complete, the other data channels cannot be considered secure. As far as I understand, the nice thing about `negotiated: true` is that we don't need to set up callbacks because of parties can use the `createDataChannel` API which will return the data channel to use for negotiating. Together with a timeout in case of a misbehaving peer, this should give us certainty that we are not processing data until the handshake is finished.
> 
> @thomaseizinger we can process the datachannels only after the handshake succeeds, but setting the callbacks after the handshake is complete causes the race condition.

Okay, this sounds like an implementation detail to me then. What I think is important is that the application layer on top of the transports cannot use the channels up until the handshake is complete. Everything else would be a leaky abstraction IMO. Without going to much into detail: In `rust-libp2p`, applications are written as plugins (`NetworkBehaviour`) that can open connections and substreams on those connections. A plugin's connection handler will only be instantiated once the handshake is complete and thus does not have access to a connection prior to it.

I am not sure to what degree we need to include this in the specification but it does seem to me that this is the first case in which access to "unsafe" channels is possible. Other combinations like TCP + noise + yamux or QUIC don't have these problems.

Reply:
> That looks like a promising idea, I like it! We could leave it empty for normal libp2p streams, and only set it to something non-empty for our handshake stream.

So taking this idea further, upon connection setup, an application would register a listener for data channels and wait until it receives the first one that says "handshake". All others will be either discarded or buffered until the handshake is complete?

Reply:
> Other combinations like TCP + noise + yamux or QUIC don't have these problems.

They do. In principle, you can send 0.5-RTT data wherever TLS 1.3 is used (that means when using TLS 1.3 / TCP, and when using QUIC). See #454. We're _currently_ not making use of this, but we will start doing so in the coming months in go-libp2p, probably.

> Okay, this sounds like an implementation detail to me then. What I think is important is that the application layer on top of the transports cannot use the channels up until the handshake is complete. Everything else would be a leaky abstraction IMO. Without going to much into detail: In rust-libp2p, applications are written as plugins (NetworkBehaviour) that can open connections and substreams on those connections. A plugin's connection handler will only be instantiated once the handshake is complete and thus does not have access to a connection prior to it.

I disagree, see https://github.com/libp2p/specs/pull/412#discussion_r978463454. Just because rust-libp2p has abstractions that don't allow this kind of optimization doesn't mean we need to forbid other implementations to optimize the handshake latency.

Reply:
> > Other combinations like TCP + noise + yamux or QUIC don't have these problems.
> 
> They do. In principle, you can send 0.5-RTT data wherever TLS 1.3 is used (that means when using TLS 1.3 / TCP, and when using QUIC). See #454. We're _currently_ not making use of this, but we will start doing so in the coming months in go-libp2p, probably.
> 
> > Okay, this sounds like an implementation detail to me then. What I think is important is that the application layer on top of the transports cannot use the channels up until the handshake is complete. Everything else would be a leaky abstraction IMO. Without going to much into detail: In rust-libp2p, applications are written as plugins (NetworkBehaviour) that can open connections and substreams on those connections. A plugin's connection handler will only be instantiated once the handshake is complete and thus does not have access to a connection prior to it.
> 
> I disagree, see [#412 (comment)](https://github.com/libp2p/specs/pull/412#discussion_r978463454). Just because rust-libp2p has abstractions that don't allow this kind of optimization doesn't mean we need to forbid other implementations to optimize the handshake latency.

I didn't see it as an optimization actually but it is an interesting take. It does however tie in well with what you wrote above: In today's libp2p, there is no concept of an "unauthenticated" or "unsafe" stream. It would be cool to have as an optimization for sure but I don't think it belongs into this spec.

Rather, I think it would be better to tackle that as a separate issue and see, what a consistent API / usage across various transports looks like. As you said above, care needs to be taken in what is sent over such a stream. In the interest of not bloating this spec too much, it seems better to just not allow these streams for now and think about the topic separately later?

Reply:
The cool thing about using 0.5-RTT data in TLS 1.3 is that for the receiver, it's indistinguishable from data sent after the handshake: 
<img width="786" alt="image" src="https://user-images.githubusercontent.com/1478487/191943244-b990f6b2-51c4-41b7-a065-d50a85ee704d.png">

The 0.5-RTT data arrives in the same flight as the server's last flight of the handshake, so unless there's packet loss, this will practically be the same time.

>  In today's libp2p, there is no concept of an "unauthenticated" or "unsafe" stream. It would be cool to have as an optimization for sure but I don't think it belongs into this spec.

I don't think we need to do anything in this spec, other than _not_ mandating to wait until handshake completion before accepting streams (while keeping in mind that only when the handshake completes the peer is authenticated). Implementations that don't care that much about optimizing things can then block, whereas implementations that do can implement those optimizations.

Reply:
> I don't think we need to do anything in this spec, other than _not_ mandating to wait until handshake completion before accepting streams (while keeping in mind that only when the handshake completes the peer is authenticated).

Fair. Do you agree that it is worth a mention to make implementations aware of the topic?

Reply:
Yes, definitely. What about:
> Implementations MAY open streams before completion of the handshake. Applications MUST take special care what application data they send, since at this point the peer is not yet authenticated. Similarly, the receiving side MAY accept streams before completion of the handshake.

Reply:
Sounds good to me, thanks for writing it out!

Reply:
666c9f4 requires the Noise handshake channel to use `id: 0` and `negotiated: true`. @ckousik @melekes let me know if that works for both of you.

I don't yet understand the benefit of using the `protocol` property over a fixed stream ID. In case you feel strongly about that, please expand on the rational. As far as I can tell, we have enough upgrade mechanism already (e.g. protocol name, Protobuf framing, ...) to evolve the libp2p WebRTC protocol in the future.

> Yes, definitely. What about:
> 
> > Implementations MAY open streams before completion of the handshake. Applications MUST take special care what application data they send, since at this point the peer is not yet authenticated. Similarly, the receiving side MAY accept streams before completion of the handshake.

Added with ae5b0d8. Thanks @marten-seemann.

Reply:
Approval from @ckousik and @melekes out of band. Will resolve here. Thanks everyone for the input.